pkg/ddevapp/config.go

```diff
@@ -1134,6 +1134,21 @@ ARG DDEV_PHP_VERSION
 ARG DDEV_DATABASE
 RUN (groupadd --gid $gid "$username" || groupadd "$username" || true) && (useradd  -l -m -s "/bin/bash" --gid "$username" --comment '' --uid $uid "$username" || useradd  -l -m -s "/bin/bash" --gid "$username" --comment '' "$username" || useradd  -l -m -s "/bin/bash" --gid "$gid" --comment '' "$username" || useradd -l -m -s "/bin/bash" --comment '' $username )
 `
+	// If our PHP version is not already provided in the ddev-webserver, add it now
+	if strings.Contains(fullpath, "webimageBuild") {
+		if _, ok := nodeps.PreinstalledPHPVersions[app.PHPVersion]; !ok {
+			contents = contents + fmt.Sprintf(`
+### DDEV-injected addition of not-preinstalled PHP version
+RUN /usr/local/bin/install_php_extensions.sh "php%s" "${TARGETPLATFORM#linux/}"
+RUN update-alternatives --set php /usr/bin/php%s
```

Comment:
I have seen the `composer update` failures with php5.6... but composer just isn't going to work anyway is it with 5.6? My assumption was that it was OK, just an artifact of ancient PHP. I don't think (any version of) composer is going to be usable with 5.6 is it?

Reply:
In short, Composer uses PHP, if the current PHP is not compatible with the current Composer (which is always latest stable in our Docker image), it cannot downgrade itself.

---

Composer 2.2 works fine with PHP 5.6, 7.0, 7.1

The problem is not whether some PHP can use Composer or not: by default we install the latest stable version of Composer and then use the `composer self-update` feature, but Composer itself depends on PHP, and if the current Composer binary from which the upgrade/downgrade is done is not compatible with the current PHP binary, it will not be able to run this upgrade/downgrade.

If someone is using PHP 5.6, 7.0, 7.1, DDEV will try to install Composer every `ddev start` because there is a warning, the warning from `log-stderr.sh` doesn't show the details, because they are written to stdout:

```
$ ddev config --php-version=7.1 --composer-version=2.2

$ ddev start
...
Warning: command 'composer self-update --2.2' run as 'root' failed with exit code 1:
Command 'composer self-update --2.2' didn't return stderr output
...

$ ddev exec composer self-update --2.2
Composer 2.3.0 dropped support for PHP <7.2.5 and you are running 7.1.33-65+0~20241104.97+debian12~1.gbpf30d7c, please upgrade PHP or use Composer 2.2 LTS via "composer self-update --2.2". Aborting.
Failed to execute command composer self-update --2.2: exit status 1
```

The same config works with DDEV v1.23.5.

Reply:
Thanks so much for the careful thinking and testing.  My interpretation of your suggestion is in https://github.com/ddev/ddev/pull/6691/commits/0acdcd11c02bf0a8c2bd282fc12b81de3851e642

It does remove the composer warning. I think it's likely correct and non-intrusive.